### PR TITLE
fix: cypress/react18 rerender

### PR DESCRIPTION
**Pull request by hiebj**
- Closes [issue-23357](https://github.com/cypress-io/cypress/issues/23357)

### User facing changelog
Fix rerender in @cypress/react18

### Additional details
The `rerender()` function returned by `mount()` creates a new `ReactDOM.Root` every time it is invoked.  This results in state being lost between rerenders, and all lifecycle hooks being rerun.

This PR fixes that by only creating a new root if one has not been created yet, or the previous one has been cleaned up.

### Steps to test
I ran the tests from the neighboring `react` project against `react18`. I have elected not to add tests to the `react18` package as it does not currently have them and that would make this PR much more substantial. I suggest finding some way to run the `react` tests on the `react18` project rather than copying or re-implementing the tests.

### How has the user experience changed?
`rerender()` works now.

### PR Tasks

- [ na ] Have tests been added/updated? **SEE Steps to test ABOVE**
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ na ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ na ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by lmiller1990 — COMMENTED** [review]
Thanks for this, we definitely need the tests but someone who knows the code base well can probably pick it up.

For whoever does, here's what we should do:

We have `react18` in `system-tests`: https://github.com/cypress-io/cypress/blob/develop/system-tests/projects/react18/package.json#L11 which has `"projectFixtureDirectory": "react"` inside `package.json. This means when we run the system test, it inherits all the code from https://github.com/cypress-io/cypress/tree/develop/system-tests/project-fixtures/react (including the tests).

So, you could add your test in [here](https://github.com/cypress-io/cypress/tree/develop/system-tests/project-fixtures/react), then all react projects get it. You'll need to update a ton of snapshots, with `cd system-tests && SNAPSHOT_UPDATE=1 yarn test component_testing_spec`.

That should give us the coverage we need across all react versions.

cc @rockindahizzy, I think you worked on React 18 so this might be relevant to you. Alternatively if @rachelruderman would like to work on the system tests with either of us, that might be a good way to share how we test the `cy.mount` adapters for Component Testing (or in this case, how we missed coverage for re-rendering).

**Comment by hiebj** [comment]
I can take a shot at implementing the tests the way you describe if one of your gurus doesn't get to it first.

**Comment by lmiller1990** [comment]
Hi @hiebj I will try to get this tested and merged up by the end of the week - sorry for the delay.

**Comment by lmiller1990** [comment]
Hi @hiebj I will try to get this tested and merged up by the end of the week - sorry for the delay.

**Comment by hiebj** [comment]
@lmiller1990 No problem. I was considering taking a shot at it myself but I have no idea which snapshots to update. I tried running it with * and of course it took forever and some of the tests failed for reasons I did not understand. I was going to come back to it in the next day or two, but if you are on it I will defer.

**Comment by lmiller1990** [comment]
> @lmiller1990 No problem. I was considering taking a shot at it myself but I have no idea which snapshots to update. I tried running it with * and of course it took forever and some of the tests failed for reasons I did not understand. I was going to come back to it in the next day or two, but if you are on it I will defer.

All good... I will write the process I follow to update everything as well, for future reference. Running the entire suite locally takes hours and hours, don't try that 😓 it's massively parallelized on CI so it takes ~20 minutes.

File: system-tests/project-fixtures/react/vite.config.js [hunk 1]
```diff
@@ -1,5 +1,11 @@
 const {defineConfig} = require('vite')
 
 module.exports = defineConfig({
+  resolve: {
```
Comment (astone123) [inline system-tests/project-fixtures/react/vite.config.js:4]: I had to add aliases for these packages, otherwise the React 17 tests that use Vite would fail when I called `React.useState` in my test.

File: system-tests/projects/react17/package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "dependencies": {
     "react": "17.0.2",
-    "react-dom": " 17.0.2"
```
Comment (astone123) [inline system-tests/projects/react17/package.json:4]: Drive-by fix. Not sure why there was a space here

**Review by ZachJW34 — APPROVED** [review]
Looks great! `unit-tests` is failing but not due to your changes https://github.com/cypress-io/cypress/issues/23623

LGTM!

**Review by astone123 — APPROVED** [review]

**Comment by astone123** [comment]
This is ready to merge, just trying to get the CI green

**Review by lmiller1990 — APPROVED** [review]
